<img src='https://i.imgur.com/RDAD11M.png' width = '200' align = 'right'>

## *DATA SCIENCE / SECTION 1 / SPRINT 2 / NOTE 2*

# 📝 Assignment

# 카이제곱검정


## 1. 공공데이터 조작 연습. 

국가 통계 포털에서 아래 이미지를 참조하여

<https://kosis.kr/statisticsList/statisticsListIndex.do?menuId=M_01_01&vwcd=MT_ZTITLE&parmTabId=M_01_01#SelectStatsBoxDiv>

**2020년 8월**에 해당하는 규모별 미분양현황 에 대한 데이터셋을 생성하세요. 

<img src='https://i.imgur.com/aZc4UJO.png' width = 500>

- `60m이하`, `60~85m`, `85m초과`의 3개 규모와 (column)
- `서울`, `대전`, `대구`, `부산` 4개의 지역을 포함해야합니다. (row)
- 데이터가 없는 경우는 0으로 처리하세요.

이후 데이터셋을 colab으로 불러오세요. 이때 변수의 이름은 `df`를 사용합니다.


In [421]:
from google.colab import drive
drive.mount('/content/drive')

col = ['~60m2', '60~85m2', '85m2~']
row = ['서울', '대전', '대구', '부산']

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [422]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [438]:
import pandas as pd
df = pd.read_excel('/content/drive/MyDrive/미분양 8월.xlsx')
df = df.fillna(0)

# 확인
df

규모(1)       60㎡이하60〜85㎡60㎡이하60〜85㎡85㎡초과60㎡이하60〜85㎡60㎡이하60〜...
2020. 08                                                 3917
dtype: object

## 2. 지역에 대해서 one-sample chi-square test를 실행, 해당 결과를 `chi1`에 저장 후 설명해보세요.

예시) 만약 **9월달 데이터**를 기준으로 한다면
```python
 [0+0+0+52+2+0, 0+0+0+590+665+142, 0+0+0+113+1061+42, 0+0+0+772+1+0]
```
을 비교 하게 될 것입니다.




In [424]:
x= df['2020. 08']

In [440]:
from scipy.stats import chisquare
from scipy.stats import chi2_contingency
#귀무가설 = 지역별로 미분양의 분포는 같다
#대립가설 = 지역별로 미분양의 분포는 다르다

obs = [x[0:5].sum(), x[5:10].sum(),x[10:15].sum(), x[15:20].sum()]
obs
chi1 = stats.chisquare(obs)

# 값 확인
chi1

# p-value 값이 0(>0.05)이므로 귀무가설을 기각할 수 없다
# 즉, 통계적으로 지역별 미분양의 분포가 비슷하다

Power_divergenceResult(statistic=1564.4572376818994, pvalue=0.0)

## 3. 지역과 규모에 대해서 two-sample chi-square test를 실행, 해당 결과를 `chi2`에 저장 후 설명해보세요.

예시) **9월달 데이터**를 기준으로 한다면

| |-60 | 60-85 | 85- |
|:-:|:-:|:-:|:-:|
|서울|52|2|0|
|대전|772|1|0|
|대구|113|1061|42|
|부산|590|665|142|

에 대해서 검정해야 할 겁니다.




In [426]:
y = x.values.tolist()

In [427]:
seoul = y[2:5]
busan = y[7:10]
daegu = y[12:15]
daejeon = y[17:20]
rename = ['서울', '부산','대구','대전']
region = pd.DataFrame([seoul,busan,daegu,daejeon],index=rename,columns=['-60','60-85','85-'])
region

-60  60-85  85-
서울   54      2    0
부산  577    735  142
대구  143   1437   44
대전  782      1    0

In [442]:
from scipy import stats
# two sample Chi-squared test
chi2 = stats.chi2_contingency(region)


#귀무가설 = 미분양 분포에 대해 지역과 규모는 관련이 있다
#대립가설 = 지역별로 미분양의 분포는 다르다

chi2 = stats.chi2_contingency(region)
#결과
chi2
# p값이 0(>0.05)이므로 귀무가설 기각하지 않으며 통계적으로 지역과 규모는 관련이 있다

(2064.576731417199, 0.0, 6, array([[ 22.24559612,  31.09522594,   2.65917794],
        [577.59101353, 807.36533061,  69.04365586],
        [645.12228746, 901.76155221,  77.11616033],
        [311.04110288, 434.77789124,  37.18100587]]))

### 4. 2번에 대해서 NumPy 를 사용하여 (Scipy를 사용하지 않고) $\chi^2$ test 시행 후 2번의 결과와 비교해보세요. 

- `obs`, `exp`, `chi`라는 변수를 사용해야합니다.

이 후 `chi2_contingency()`의 결과물과 비교해보세요

In [429]:
import numpy as np

In [430]:
import numpy as np
from scipy.stats import chisquare  


obs = np.array([[x[0:5].sum(), x[5:10].sum(),x[10:15].sum(), x[15:20].sum()]])
obs

array([[  56, 1454, 1624,  783]])

In [431]:
total_obs = np.sum(obs)
total_obs

3917

In [443]:
exp = np.array([56*total_obs, 1454*total_obs, 1624*total_obs, 783*total_obs])/total_obs
exp = exp/total_obs
exp

array([0.01429666, 0.37120245, 0.41460301, 0.19989788])

In [433]:
squared = np.power(obs - exp,2) #obs 관찰값 exp 예상값
squared

array([[   3134.39897897, 2113036.68106417, 2636029.54131103,
         612775.99987747]])

In [447]:
chi = round(np.sum(squared/exp),3)
#결과
chi

15335056.0

In [446]:
#2번의 카이제곱 통계치와 값이 다르다
#2번 통계치 = 1564.457
#4번 통계치 = 15335056

chi2_contingency(obs)
#chi2_contingency통계치 = 0

(0.0, 1.0, 0, array([  56., 1454., 1624.,  783.]))

## 🔥 도전 과제

아래 세가지를 다하면 됩니다.

### 1. Function

4번에서 사용한 one sample chisquare test를 함수의 형태로 변경하세요.

In [436]:
v1 = [18,22,20,15,23,22]
v2 = [5,23,26,19,24,23]

def myChisq(value):
    return None# 내용을 채우세요.

myChisq(v1) 

# chisquare값과 pvalue를 출력해야합니다.

myChisq(v2)

### 2. ANOVA

아래 링크를 참조하여 ANOVA 에 대한 글을 읽고

<https://partrita.github.io/posts/ANOVA-python/>

다음 `4개 그룹에 대해서 평균의 차이가 있는지`에 대한 가설 검정을 시행하세요.

A : `38 33 35 92 76 97 88 41 11  9`

B : `18 52 62 48 30 40 87 12 97 82`

C :  `28 90  5 49 66 73 96 80  4 17`

D : ` 8 99  4 12  7 64 18 10  9 20`



## 3. Sign Test

아래 링크를 참조하여 Sign Test에 대한 글을 읽고

<https://en.wikipedia.org/wiki/Sign_test>

다음 데이터에서 두 그룹간에 차이가 있다에 대한 Sign Test의 과정과 결과를 설명하세요

|A|B|
|:-:|:-:|
|142|138|
|140|136|
|144|147|	
|144|139|	
|142|143|	
|146|141|	
|149|143|	
|150|145|	
|142|136|	
|148|146|	

